### PR TITLE
make --socket-path pull ETHERSYNC_SOCKET from env via clap

### PR DESCRIPTION
**Pull request by cosmicexplorer**
# Problem
It's much easier to set `ETHERSYNC_SOCKET` via the environment for many use cases instead of modifying the command line, as noted by @zormit in https://github.com/ethersync/ethersync/pull/112#issuecomment-2287008800. We currently do this by hand in the nvim plugin: https://github.com/ethersync/ethersync/blob/a41304aa0543077bcf8908d912dabf8bc91069cf/vim-plugin/plugin/ethersync.lua#L69-L72

We could do the same for the emacs plugin in #112, but it might be nice for the `ethersync` binary to support this itself.

# Solution
- Add `"env"` feature to our `clap` dependency.
- Set [`Arg::env()`](https://docs.rs/clap/latest/clap/struct.Arg.html#method.env) for the global `socket_path: PathBuf` argument.
    - A command-line `--socket-path` argument overrides this value; the env var is only read if the argument is not provided.
    - *Note that the daemon is now also affected by `ETHERSYNC_SOCKET` in addition to the client.*
- Remove code in nvim plugin to translate the `ETHERSYNC_SOCKET` env var into `--socket-path`.

# Result
All editor plugins (and the fuzzer) can expect to be able to set the `ETHERSYNC_SOCKET` environment variable as a separate way to parameterize the ethersync client, and users can set `ETHERSYNC_SOCKET` in their environment when invoking the daemon.

### CODE REVIEW

File: daemon/src/main.rs [hunk 1]
```diff
@@ -19,10 +19,14 @@ struct Cli {
...
     /// Path to the Unix domain socket to use for communication between daemon and editors.
-    #[arg(short, long, global = true, default_value = DEFAULT_SOCKET_PATH)]
+    #[arg(
+      short, long, global = true,
+      default_value = DEFAULT_SOCKET_PATH,
+      env = "ETHERSYNC_SOCKET",
+    )]
     socket_path: PathBuf,
     /// Enable verbose debug output.
-    #[arg(short, long, global = true, action)]
```
Comment (cosmicexplorer) [inline daemon/src/main.rs:25]: This doesn't seem to change behavior, as can be seen by the output of `ethersync --help`. If the intent was to allow `--debug=true`, I think you would need something like `action(clap::ArgAction::Set), default_value = "false"`, but I'm having difficulty figuring out how to make that work.

If the intent was just to make a simple boolean flag `-d`/`--debug`, I think this `action` argument is redundant. Let me know if I'm mistaken!

Comment (blinry) [reply daemon/src/main.rs:25]: You seem right! Thanks for removing the redundancy! :)

File: daemon/src/main.rs [hunk 2]
```diff
@@ -69,6 +73,7 @@ async fn main() -> Result<()> {
     logging::initialize(cli.debug);
 
     let socket_path = cli.socket_path;
+    info!("Using socket path {}", socket_path.display());
```
Comment (blinry) [inline daemon/src/main.rs]: I noticed that `ethersync client` will also print this info log. While this is helpful for humans, because editor plugins use the command to "speak" JSON-RPC, I think it'd be good to not show it for the `client` subcommand.

For the daemon, maybe we could print it only if read from the environment variable? (That's the "surprising" case.) Not sure if it's easy to figure that out after Clap has run.

File: vim-plugin/plugin/ethersync.lua [hunk 3]
```diff
@@ -66,11 +66,6 @@ local function connect()
 
     local params = { "client" }
 
-    local socket_path = os.getenv("ETHERSYNC_SOCKET")
-    if socket_path then
-        table.insert(params, "--socket-path=" .. socket_path)
-    end
-
```
Comment (blinry) [inline vim-plugin/plugin/ethersync.lua:73]: I really like that we can get rid of this! :) Props to you for noticing that. Takes away a bit of complexity in the editor plugins.

File: daemon/src/actors.rs [hunk 4]
```diff
@@ -25,6 +25,7 @@ impl Neovim {
     /// Will panic if Neovim cannot be started or if the file cannot be opened.
     pub async fn new(file_path: PathBuf) -> Self {
         let handler = Dummy::new();
+        /* TODO: consider a process to bootstrap nvim for testing in CI. */
```
Comment (blinry) [inline daemon/src/actors.rs]: There is a (slightly involved) way: We provide a Nix flake that'll make a Neovim with Ethersync plugin available.

You can try it using `nix run github:ethersync/ethersync#neovim`, or put it in your PATH using `nix shell github:ethersync/ethersync#neovim`.

But it's true that it'd be nice to allow people to just run all tests without having to think about this.

File: daemon/src/bin/fuzzer.rs [hunk 5]
```diff
@@ -33,6 +33,8 @@ fn initialize_project() -> (temp_dir::TempDir, PathBuf) {
     (dir, file)
 }
 
+/* TODO: consider moving fuzzer to separate crate to avoid adding test code to ethersync binary. */
+/* TODO: consider using cargo-fuzz and exposing a stdin interface to avoid network overhead. */
```
Comment (blinry) [inline daemon/src/bin/fuzzer.rs]: I think it's nice that we can test the whole "stack" here – it's our "integration test", if you will. So testing this through the network is part of the point here! :)

File: daemon/src/bin/fuzzer.rs [hunk 6]
```diff
@@ -33,6 +33,8 @@ fn initialize_project() -> (temp_dir::TempDir, PathBuf) {
     (dir, file)
 }
 
+/* TODO: consider moving fuzzer to separate crate to avoid adding test code to ethersync binary. */
```
Comment (blinry) [inline daemon/src/bin/fuzzer.rs]: Maybe an actual [integration test](https://doc.rust-lang.org/rust-by-example/testing/integration_testing.html) would be the correct place for this! This would also avoid the problem of having a "fuzzer" binary in our output if you run `cargo build`. I created #128 to track this. Feel free to remove this TODO :)

**Review by blinry — CHANGES_REQUESTED** [review]
Thanks for the thoughtful PR! Very cool that clap has a built-in feature for this!

File: daemon/src/main.rs (hunk 1 above)
Comment (zormit) [reply daemon/src/main.rs:25]: Good catch. The intent was the latter, I'm not sure why we added it in the first place if it does nothing. I think a possible alternative is to make it a count by repeating it, like [here](https://docs.rs/clap/latest/clap/_derive/_tutorial/chapter_0/index.html).

File: daemon/src/main.rs (hunk 2 above)
Comment (cosmicexplorer) [reply daemon/src/main.rs]: Not quite *after* clap has run, but rather as part of it, we can extract `ArgMatches`! I've done this (only for the daemon subcommand). Let me know what you think ^_^!

File: vim-plugin/plugin/ethersync.lua (hunk 3 above)
Comment (cosmicexplorer) [reply vim-plugin/plugin/ethersync.lua:73]: ^\_^\_^\_^\_^\_^\_^\_^\_^!!!!

File: daemon/src/actors.rs (hunk 4 above)
Comment (cosmicexplorer) [reply daemon/src/actors.rs]: I also didn't notice that your github actions workflow installs it from a tarball!! I first actually left a TODO mentioning [`spack-rs`](https://docs.rs/spack-rs) ([repo](https://github.com/cosmicexplorer/spack-rs)), my cargo wrapper for [`spack`](https://github.com/spack/spack) which can bootstrap anything spack can, including neovim. However, there's some [optimization for bootstrapping](https://github.com/spack/spack/discussions/44714) I have yet to do, and `spack-rs` in particular is for linking libraries--no reason to use it  over `spack` itself for executables.

spack is a really interesting and useful project (build any codebase on any platform without any installation requirements, install multiple versions of each package, [really cool dependency "spec" syntax](https://spack.readthedocs.io/en/latest/basic_usage.html#specs-dependencies)). You can think of it as similar to nix but with a dependency resolver and no install requirements. But since I realized we're not linking any libraries and nvim is rather easy to install, I can't immediately see where spack would be useful for ethersync at this moment.

I have removed this comment, because I didn't realize CI already bootstraps nvim. If `ethersync` ever exports a library instead of just a binary, or it wants to depend on any C/C++ library, then it may become relevant.

Please feel free to ignore this comment, I like spack quite a lot but I think it's not useful right now. ^_^

File: daemon/src/bin/fuzzer.rs (hunk 6 above)
Comment (cosmicexplorer) [reply daemon/src/bin/fuzzer.rs]: I ended up spending some time doing almost exactly this, but had created a subcrate instead of an integration test (https://github.com/ethersync/ethersync/compare/main...cosmicexplorer:use-cargo-test-for-integration-tests?expand=1). I was literally about to post a separate PR for it when I saw your review of this one -- I'll update that PR to make it an integration test instead!

File: daemon/src/bin/fuzzer.rs (hunk 5 above)
Comment (cosmicexplorer) [reply daemon/src/bin/fuzzer.rs]: So the reason I made this comment is because I believe the term "fuzz" has a relatively specific meaning in cargo, including special support with `cargo fuzz`: https://rust-fuzz.github.io/book/cargo-fuzz.html. We have a separate subcrate for fuzzing in the `zip` crate (https://github.com/zip-rs/zip2/blob/af33ed343da5769e0587d638e435d67151990f10/fuzz/Cargo.toml#L8-L9), because it often requires different dependencies, and `cargo fuzz` links the binary against clang's ASan (which causes errors if it's in the same workspace as our `zip` library). I have contributed some changes to `zip`'s fuzzing -- I would *love* to do the same here (in a separate change).

In https://github.com/ethersync/ethersync/compare/main...cosmicexplorer:use-cargo-test-for-integration-tests?expand=1 (which I am going to move into an integration test before posting as a PR), I rename the binary from `fuzzer.rs`, because I thought it would be useful to make clear it's not using the *cargo fuzz* framework. However, I think that change is actually not necessary, since it definitely is still performing a sort of fuzzing, so I will change its name back before posting the integration test PR.

File: daemon/src/bin/fuzzer.rs (hunk 6 above)
Comment (cosmicexplorer) [reply daemon/src/bin/fuzzer.rs]: Done! https://github.com/ethersync/ethersync/pull/129

File: daemon/src/main.rs (hunk 1 above)
Comment (cosmicexplorer) [reply daemon/src/main.rs:25]: I happened to have read through most of `clap`'s documentation literally two days ago for the `zip` crate (zip-rs/zip2#235), so it was a fortunate coincidence -- `clap` is very powerful, perhaps a little *too* powerful.

File: daemon/src/bin/fuzzer.rs (hunk 5 above)
Comment (cosmicexplorer) [reply daemon/src/bin/fuzzer.rs]: Broken out into #130!

File: daemon/src/bin/fuzzer.rs
Comment (zormit) [inline daemon/src/bin/fuzzer.rs:1]: (Leaving another single comment while looking at your PRs)

It seems to me this mode/permission change should not be part of this PR?

**Review by blinry — DISMISSED** [review]
This seems really good to me! :sparkles:

The last thing I'd request is undoing the mode change on fuzzer.rs spotted by @zormit :)

After that, it'd be great if you could squash the small "fixup" commits into the others, and then we can merge this! :tada:

File: daemon/src/bin/fuzzer.rs
Comment (cosmicexplorer) [reply daemon/src/bin/fuzzer.rs:1]: Thank you!! I'm very sorry, there is an issue with my editor configuration that happened a couple months ago that keeps causing this, I believe related to magit somehow but not sure. Haven't fixed it yet. Thanks so much, will check this in the future.

Comment (cosmicexplorer) [reply daemon/src/bin/fuzzer.rs:1]: Just set a pre-commit hook to make sure to avoid it in the future ^_^!

**Review by zormit — APPROVED** [review]

**Comment by zormit** [comment]
Thanks for this, @cosmicexplorer! =)
